### PR TITLE
Update PMD to version 5.4.1 and enable addition validation rules.

### PR DESCRIPTION
**Pull request by Happy-Neko**
Fix for #540.
Update PMD to version 5.4.1 and enable addition validation rules.

### CODE REVIEW

**Comment by Happy-Neko** [comment]
@davvd please proceed with code review.

**Comment by krzyk** [comment]
@yegor256 @davvd is really slow, especially on weekends, here we have a PR that is waiting for 3 days without a reviewer, other PR and issue assignment are equally slow.

**Comment by davvd** [comment]
@happy-neko Thanks, I will find someone to review this PR soon

**Comment by davvd** [comment]
@pinaf it's yours, please review

**Comment by krzyk** [comment]
@Happy-Neko you have to merge master changes

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java [hunk 1]
```diff
@@ -315,4 +320,109 @@ public void acceptsCallToConstructorInConstructor()
...
     }
+
+    /**
+     * PMDValidator forbids non public clone methods (PMD rule
+     * rulesets/java/clone.xml/CloneMethodMustBePublic).
+     * @throws Exception If something wrong happens inside.
+     */
+    @Test
+    public void forbidsNonPublicCloneMethod() throws Exception {
+        final String file = "CloneMethodMustBePublic.java";
```
Comment (pinaf) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @Happy-Neko we can inline this variable

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java [hunk 2]
```diff
@@ -315,4 +320,109 @@ public void acceptsCallToConstructorInConstructor()
...
+    /**
+     * PMDValidator forbids clone methods with return type not matching class
+     * name (PMD rule
+     * rulesets/java/clone.xml/CloneMethodReturnTypeMustMatchClassName).
+     * @throws Exception If something wrong happens inside.
+     */
+    @Test
+    public void forbidsCloneMethodReturnTypeNotMatchingClassName()
+        throws Exception {
+        final String file = "CloneMethodReturnTypeMustMatchClassName.java";
```
Comment (pinaf) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @Happy-Neko we can inline this variable

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java [hunk 3]
```diff
@@ -315,4 +320,109 @@ public void acceptsCallToConstructorInConstructor()
...
+
+    /**
+     * PMDValidator forbids ternary operators that can be simplified (PMD rule
+     * rulesets/java/basic.xml/SimplifiedTernary).
+     * @throws Exception If something wrong happens inside.
+     */
+    @Test
+    public void forbidsNonSimplifiedTernaryOperators()
+        throws Exception {
+        final String file = "SimplifiedTernary.java";
```
Comment (pinaf) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @Happy-Neko we can inline this variable

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java [hunk 4]
```diff
@@ -315,4 +320,109 @@ public void acceptsCallToConstructorInConstructor()
...
+
+    /**
+     * PMDValidator forbids non caching getInstance methods (PMD rule
+     * rulesets/java/design.xml/SingletonClassReturningNewInstance).
+     * @throws Exception If something wrong happens inside.
+     */
+    @Test
+    public void forbidsSingletonClassesReturningNewInstance()
+        throws Exception {
+        final String file = "SingletonClassReturningNewInstance.java";
```
Comment (pinaf) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @Happy-Neko we can inline this variable

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java [hunk 5]
```diff
@@ -315,4 +320,109 @@ public void acceptsCallToConstructorInConstructor()
...
+
+    /**
+     * PMDValidator forbids singleton with multiple getInstance methods
+     * (PMD rule rulesets/java/design.xml/SingleMethodSingleton).
+     * @throws Exception If something wrong happens inside.
+     */
+    @Test
+    public void forbidsSingletonWithMultipleGetInstanceMethods()
+        throws Exception {
+        final String file = "SingleMethodSingleton.java";
```
Comment (pinaf) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @Happy-Neko we can inline this variable

**Comment by pinaf** [comment]
@Happy-Neko what about the rename mentioned at the end of the issue description?

**Comment by Happy-Neko** [comment]
@pinaf thanks for the code review. Changes committed.
> @Happy-Neko what about the rename mentioned at the end of the issue description?

I didn't find `UseSingleton` usage in qulice code base.

**Comment by pinaf** [comment]
@rultor merge

**Comment by Happy-Neko** [comment]
@krzyk

> @Happy-Neko you have to merge master changes

Done. Please confirm merge.

**Comment by krzyk** [comment]
@Happy-Neko Please remove tests for singletons, they are a bad design anyway, if they are enabled in PMD it is ok, but I wouldn't die for it. I would just remove all singletons, so no point in checking if they are written slightly better :)

**Comment by Happy-Neko** [comment]
@krzyk ok. Done.

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by davvd** [comment]
@pinaf paid, thanks, added **22 mins** to your account, payment `AP-15X66246PD060252J`, 48 hours and 18 mins was spent. you're getting extra minutes here (c=7). +22 to your rating, your total score is [+5887](http://www.netbout.com/b/35128?open=rating)

**Comment by davvd** [comment]
@rultor deploy now pls

**Pull request merged**